Plugin generator

KrauseFx opened the pull request.

Commit:
Initial work on plugin generator

Commit:
Add more structure and tests around plugin generation

Commit:
Add tests to specifically test validation/fixing behavior

Commit:
Minimal spec for PluginGeneratoer

Commit:
Create spec lets

Commit:
One-time initialization in plugin generator spec

Commit:
Begin templating files as part of plugin generation

Test the correctness of the gemspec by evaluating it and
checking its values

Commit:
Fix hound warnings about exists?

KrauseFx commented on fastlane/lib/fastlane.rb:
```diff
@@ -19,6 +19,7 @@
 require 'fastlane/documentation/docs_generator'
 require 'fastlane/other_action'
 require 'fastlane/plugin_manager'
+require 'fastlane/plugins/plugins'
```
Looking at the directory structure it seems like we want all plugin related code in that folder, right? So we're gonna move the `plugin_manager`?

KrauseFx replied:
Or should it be specific to creating new plugins?

mfurtak replied:
Yeah, now that we have a group of classes, I think it might make sense to move the `plugin_manager` in there too.

KrauseFx replied:
Perfect, in the mean-time I actually have more classes already, which I'll be moving into the subfolder after merging in your changes.

KrauseFx commented on fastlane/lib/fastlane/commands_generator.rb:
```diff
@@ -175,6 +175,15 @@ def run
       # @!group Plugins
       #####################################################
 
+      command :create_plugin do |c|
```
In the docs I called the command `new_plugin`: https://github.com/fastlane/fastlane/pull/4722/files#diff-cf9ca7bd4b992dfc6d9b6cdb37a890c0R54

We can do either way, however similar tools like danger use `new` instead of `create`: https://github.com/danger/danger/blob/18d5875ab2da882b068e7d2356d1f444a82dfaed/lib/danger/commands/new_plugin.rb#L4

mfurtak replied:
`new_` is fine with me!

KrauseFx commented on fastlane/lib/fastlane/plugins/plugin_generator_ui.rb:
```diff
@@ -0,0 +1,15 @@
+module Fastlane
+  class PluginGeneratorUI
```
Why do we need a new UI class instead of using the one from `fastlane_core`? We can also easily mock the `UI` class from core.

Commit:
Fix plugin_generator_ui_spec

KrauseFx commented on fastlane/lib/fastlane/plugins/plugin_info_collector.rb:
```diff
@@ -0,0 +1,67 @@
+        unless plugin_name_valid?(plugin_name)
+          fixed_name = fix_plugin_name(plugin_name)
+          if plugin_name_valid?(fixed_name)
+            plugin_name = fixed_name if @ui.confirm("Is '#{fixed_name}' OK?")
```
To be consistent with the rest of the code base, let's use `okay` instead of `OK`. When using the fastlane_core UI layer, you'd get the ` (y/n)` at the end of the question for free. We should add it, so that's it's clear what fastlane wants to hear

mfurtak replied:
'okay' ✅ 

The `PluginGeneratorUI` class is delegating to the `FastlaneCore` UI layer, so we will get that prompt behavior. 👍 We'll think some more about how much it is actually buying us.

KrauseFx commented on fastlane/lib/fastlane/plugins/plugin_info_collector.rb:
```diff
@@ -0,0 +1,67 @@
+    def collect_plugin_name
+      plugin_name = nil
+      loop do
+        plugin_name = @ui.input("What would you like to be the name of your plugin?")
```
At this point it's not clear if the plugin should be pre-fixed, right? Let's assume we build a plugin called `hockey_dsym_upload`:

The user could just enter `hockey_dsym_upload` and get the `fastlane-plugin-` prefix for free, right? If the user pre-fixes the input here, we should automatically detect that

mfurtak replied:
This method will detect if the user types in the prefix. The idea is that `plugin_name_valid?` will find the prefix and return falsey, which causes `fix_plugin_name` to run and produce a name suggestion that has the prefix removed. 👍

KrauseFx commented on fastlane/lib/fastlane/plugins/plugin_info_collector.rb:
```diff
@@ -0,0 +1,67 @@
+
+    # Consider testing this more directly?
+    def plugin_name_valid?(name)
+      /^[a-z0-9_]+$/ =~ name && !name.downcase.start_with?('fastlane_')
```
Looks great. Instead of hardcoding the prefix, you can access `PluginManager.plugin_prefix`

KrauseFx commented on fastlane/lib/fastlane/plugins/plugin_info_collector.rb:
```diff
@@ -0,0 +1,67 @@
+    end
+
+    def fix_plugin_name(name)
+      name.to_s.downcase.gsub(/[\- ]/, '_').gsub(/[^a-z0-9_]/, '').gsub(/^fastlane_/, '')
```
Similar to https://github.com/fastlane/fastlane/pull/4762/files#r64125613

KrauseFx replied:
Maybe add a comment on what this method does 👍

KrauseFx commented on fastlane/lib/fastlane/plugins/templates/README.md.erb:
````diff
@@ -0,0 +1,9 @@
+fastlane add_plugin <%= plugin_name %>
+```
+
+TODO: Write this README for real!
````
❤️

KrauseFx commented on fastlane/lib/fastlane/plugins/templates/plugin.gemspec.erb:
```diff
@@ -0,0 +1,20 @@
+# coding: utf-8
+lib = File.expand_path('../lib', __FILE__)
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+require '<%= gem_name %>/version'
```
Beautiful 👍

KrauseFx commented on fastlane/lib/fastlane/plugins/templates/plugin.gemspec.erb:
```diff
@@ -0,0 +1,20 @@
+  spec.name          = '<%= gem_name %>'
+  spec.version       = Fastlane::<%= plugin_name.capitalize %>::VERSION
+  spec.author        = '<%= author %>'
+  #spec.email         = ["fastlane-xcversion@gmail.com"]
```
Maybe we should just add our dev tools email here? Some developer might just comment it out, and then they have a valid email address where they can reach us 👍

KrauseFx commented on fastlane/lib/fastlane/plugins/templates/plugin.gemspec.erb:
```diff
@@ -0,0 +1,20 @@
+  #spec.email         = ["fastlane-xcversion@gmail.com"]
+
+  #spec.summary       = %q{Something random}
+  #spec.homepage      = "https://github.com/fastlane/fastlane-xcversion"
```
Let's point to the fastlane main repo instead 👍

KrauseFx commented on fastlane/lib/fastlane/plugins/templates/plugin.gemspec.erb:
```diff
@@ -0,0 +1,20 @@
+  #spec.files         = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
+  #spec.require_paths = ["lib"]
+
+  #spec.add_dependency 'xcode-install', '~> 1.3.0'
```
Let's add spaces after the `#`

mfurtak replied:
Agreed. This file is still under active development 👍

Commit:
Address feedback, adapt to new prefix, fix paths

* Improve ui object mocking to prevent test hangs if an
  interactive method is not properly mocked
* Fix file generation paths to be correctly start inside
  the gem name directory
* Fix gemspec evaluation during tests to account for its
  new location relative to the temp directory
* Fix plugin info collection to be aware of the changed
  fastlane plugin prefix value
* PluginInfo can generate the require_path from the gem_name
* Command name changed to 'new_plugin'

KrauseFx commented:
I didn't dive into the code, but we have to make sure the plugin's name doesn't contain a `-`, as this would cause issues when requiring them gem.

mfurtak commented:
@KrauseFx Re: https://github.com/fastlane/fastlane/pull/4762#issuecomment-220803964

Yes - https://github.com/fastlane/fastlane/pull/4762/files#diff-d9ef8d1a145e51d4dbfb0269eb42eee9R43 and https://github.com/fastlane/fastlane/pull/4762/files#diff-d9ef8d1a145e51d4dbfb0269eb42eee9R54 take care of that. 👍

mfurtak committed:
Clean up/suppress rubocop violations

Commit:
Add email collection to the plugin generator interview

Commit:
Pre-fill summary, description, homepage in gemspec; reword info collectory questions

Commit:
LICENSE creation

Commit:
Add support for collecting gemspec summary and description

Commit:
Add missing LICENSE file

Commit:
Accept plugin name during command line call

Commit:
Define the plugin action class as part of the templating

Commit:
Make the plugin gem depend on fastlane

Commit:
Make fastlane a development dependency of the plugin bundle

* Add other needed development dependencies to the plugin bundle
  template

Commit:
Fix plugin_generator_spec

Commit:
Add helper file generation to the plugin generator

Commit:
Add .rspec file to the plugin generator content

Commit:
Fix failing rubocop offense

Commit:
Add Gemfile to plugin generator content

Commit:
Add .gitignore to the plugin generator content

KrauseFx commented on fastlane/lib/fastlane/plugins/plugin_info_collector.rb:
```diff
@@ -0,0 +1,129 @@
+    # Author
+    #
+
+    def collect_author
```
Could we use the Git name as default, similar to what CocoaPods does? Or did we decide against it?

KrauseFx replied:
Same for email address

mfurtak replied:
We decided to implement it for v2.

KrauseFx commented on fastlane/lib/fastlane/plugins/plugin_info_collector.rb:
```diff
@@ -0,0 +1,129 @@
+    #
+
+    def collect_description
+      @ui.input("\nPlease enter a longer description of this fastlane plugin:")
```
Collecting just one description should be good enough, most people would be annoyed when they just finished writing the description, and then they're asked for an even longer description.

mfurtak replied:
OK, let me verify that it is only a warning if the description is missing, and if so I'll take it out.

KrauseFx commented on fastlane/lib/fastlane/plugins/templates/LICENSE.erb:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) <%= Time.now.year %> <%= author %> <<%= email %>>
```
Duplicate `<`

mfurtak replied:
That is intentional so that the line looks like 
```
Copyright (c) 2016 Michael Furtak <mfurtak@twitter.com>
```

KrauseFx replied:
Aww, sorry, missed that

KrauseFx commented on fastlane/lib/fastlane/plugins/templates/plugin.gemspec.erb:
```diff
@@ -0,0 +1,27 @@
+
+  spec.summary       = %q{<%= summary %>}
+  spec.description   = %q{<%= description %>}
+  spec.homepage      = "https://github.com/<GITHUB_USERNAME>/<%= gem_name %>"
```
https://github.com/KrauseFx/fastlane-plugin-ruby/commit/588351fdd56a6f9bac629c882ddd683f92396d61#diff-0c3ae00cec96b432904720470bc8c426R14
Could we fetch this from Git too?

KrauseFx replied:
Sorry if you discussed this already, just let me know how you decided 👍

mfurtak replied:
We'll look at it for v2.

KrauseFx replied:
Sounds good

KrauseFx commented:
This is great already, look at how light-weight everything is: https://github.com/KrauseFx/fastlane-plugin-ruby/commit/588351fdd56a6f9bac629c882ddd683f92396d61#diff-0c3ae00cec96b432904720470bc8c426R14

I just noticed that the `[plugin-name].rb` doesn't get created automatically. The content should be like this https://github.com/KrauseFx/fastlane-plugin-ruby/blob/master/lib/fastlane/plugin/ruby.rb

mfurtak commented:
Nice! Yes, the main `.rb` file still needs to be generated along with the specs.

Commit:
Add main plugin .rb file to the plugin generator content

Commit:
Update README template for plugin with badge.

Commit:
Add a success message when the plugin generator is done

Commit:
Add spec helper generation

Commit:
Add action_spec creation to the plugin generator

Commit:
Add Rakefile generation

Commit:
Remove description from PluginInfo model

Commit:
Add more details the generated README.md

Commit:
Continue improving the documentation for plugins and generator

KrauseFx commented:
@mfurtak re https://github.com/fastlane/fastlane/pull/4762#discussion_r64749320, I committed https://github.com/KrauseFx/fastlane-plugin-ruby/commit/d99dc8058594c64ac0a8f16032e19993b5536382 and the description looks good: https://rubygems.org/gems/fastlane-plugin-ruby. Also in the `fastlane search_plugins` everything is okay 👍 Just having the summary is fine

KrauseFx commented on fastlane/docs/PluginsTroubleshooting.md:
```diff
@@ -50,12 +50,12 @@ Your `Pluginfile` should look something like this
 # Autogenerated by fastlane
 
-gem 'fastlane-plugin-xcversion'
+gem 'fastlane-plugin-ruby'
```
@mfurtak Once we have more useful plugins we should add this here, since `ruby` is irrelevant for most iOS/Android devs

mfurtak replied:
@ohwutup pointed out to me that fastlane has a built-in action called xcversion. Seeing that, it wasn't clear to me how the xcversion example plugin works at all. This was just being a little defensive, just in case

KrauseFx commented on fastlane/lib/fastlane/plugins/plugin_generator.rb:
```diff
@@ -0,0 +1,109 @@
+      @ui.success "\nTo get started with using this plugin, run"
+      @ui.message "\n    fastlane add_plugin #{plugin_info.plugin_name}\n"
+      @ui.success "\nfrom a fastlane-enabled app project directory and provide the following as the path:"
+      @ui.message "\n    #{File.expand_path(plugin_info.gem_name)}\n\n"
```
The output looks like this for me:
```
Your plugin was successfully generated at fastlane-plugin-ohyeah/ 🚀

To get started with using this plugin, run

    fastlane add_plugin ohyeah

from a fastlane-enabled app project directory and provide the following as the path:

    /Users/felixkrause/Developer/fastlane/fastlane/fastlane-plugin-ohyeah
```
I think the `\n` are not necessary here

KrauseFx replied:
Maybe we should add
```ruby
@ui.message "\nOnce your plugin is ready to be published, check out the plugins guide on GitHub: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Plugins.md"
```

mfurtak replied:
We had spaced it out because we thought it highlighted the info we want you to see (the command and the path) a bit better. We can squash it if you like that better 👍

mfurtak replied:
We can add more, but the message is getting long, and I worry people won't read it.

KrauseFx commented:
The badge doesn't get copied over, investing this now

KrauseFx committed:
Renamed fastlane logo

KrauseFx commented on fastlane/lib/fastlane/plugins/templates/action.rb.erb:
```diff
@@ -0,0 +1,35 @@
+      end
+
+      def self.description
+        %q{<%= summary %>}
```
I'm not a big fan of the %q, this is very confusing to the user, I'd much rather prefer using `"` and escaping the `"`. What do you think?

KrauseFx replied:
Oh this is used in various places, I'll submit a separate commit

mfurtak replied:
Hm, trying to manually escape user input is usually a bad idea, as we have seen in other places 😅 But if you feel strongly, we can try it!

KrauseFx replied:
Is there any issue with escaping the `"` character? I really don't like the %q as it adds quite some unnecessary complexity. Also with `%q{` we also have to escape `}`, right?

mfurtak replied:
Yes, the last point is fair. I guess `%q` just requires less work for the likely case. I believe we copied the strategy from Cocoapods, for what that's worth.

I guess my point about trying to do escaping is that it is always more tricky than you initially think. For example given the user input that results in:

```
My description is either "good"\"bad"
```

Just escaping `"` leads to:

```
My description is either \"good\"\\"bad\"
```

which I believe will be broken (the double backslash escapes the backslash, not the quotes) if you place inside of a 
```
"<%= template %>"
```

mfurtak replied:
I don't mean to be obstructionist about this. I totally understand why you feel like the special syntax is confusing, so if you feel strongly that it should be changed, go for it! :smile:

KrauseFx commented on fastlane/lib/fastlane/plugins/templates/plugin.rb.erb:
```diff
@@ -0,0 +1,16 @@
+require '<%= require_path %>/version'
+
+module Fastlane
+  module <%= plugin_name.capitalize %>
```
Ah, that doesn't work, this should be `plugin_name.fastlane_class`, so it works with `_` too 👍

mfurtak replied:
Good catch!
